harness/determined/pytorch/_reducer.py

```diff
@@ -1,9 +1,176 @@
+import abc
+import logging
 from enum import Enum
-from typing import List, Optional
+from typing import Any, Callable, List, Type, Union
 
 import numpy as np
 
-import determined_common.check as check
+ReducerType = Union["Reducer", Type["MetricReducer"], Callable]
+
+
+class MetricReducer:
+    """
+    Efficiently aggregating validation metrics across a multi-slot distributed evaluation is done
+    in two steps:
+
+    1. Accumulate metrics from each batch on each slot. In the case of calculating a mean, this
+       might mean keeping a running sum and a count of metrics received.
+
+    2. Reduce metrics from each slot to calculate the final metric. In the case of calculating a
+       mean, this might mean adding up the per-slot sums and dividing the result by the per-slot
+       counts.
+
+    Example implementation and usage:
+
+    .. code:: python
+
+        class MyAvgMetricReducer(pytorch.MetricReducer):
+            def __init__(self):
+               self.sum = 0
+               self.counts = 0
+
+            def accumulate(self, metric):
+                self.sum += sum(metric)
+                self.counts += 1
+                return self.sum, self.counts
+
+            def cross_slot_reduce(self, per_slot_metrics):
+                # per_slot_metrics is a list of (sum, counts) tuples
+                # returned by the final self.accumulate() on each slot
+                sums, counts = zip(*per_slot_metrics)
+                return sum(sums) / sum(counts)
+
+        class MyPytorchTrial(pytorch.PyTorchTrial):
+            ...
+            def evaluate_batch(...):
+                ...
+                return {"my_averageable_metric": metric_tensor, ...}
+
+            def evaluation_reducer(...):
+                return {"my_averageable_metric": MyAvgMetricReducer, ...}
+
+    See also: :meth:`determined.pytorch.PyTorchTrial.evaluation_reducer`.
+    """
+
+    @abc.abstractmethod
+    def accumulate(self, metric: Any) -> Any:
+        """
+        accumulate is called for each batch in the evaluation dataset.  Batches will be distributed
```

Comment:
non-blocking: "accumulate is called for each batch in the evaluation dataset.  Batches will be distributed across slots, so accumulate will be called many times on each slot." -> "accumulate is called for each batch in the evaluation dataset, the process of which happens on each slot respectively."